server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/model/MailboxFactory.scala

```diff
@@ -0,0 +1,122 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.jmap.model
+
+import eu.timepit.refined.api.Refined
+import javax.inject.Inject
+import org.apache.james.jmap.mail.Mailbox.MailboxName
+import org.apache.james.jmap.mail._
+import org.apache.james.jmap.utils.quotas.QuotaLoader
+import org.apache.james.mailbox.model.MailboxACL.EntryKey
+import org.apache.james.mailbox.model.{MailboxCounters, MailboxId, MailboxMetaData, MailboxPath, MailboxACL => JavaMailboxACL}
+import org.apache.james.mailbox.{MailboxSession, Role, SubscriptionManager}
+
+import scala.jdk.CollectionConverters._
+import scala.jdk.OptionConverters._
+
+sealed trait MailboxConstructionOrder
+
+case class AllMailboxesItem(mailboxMetaData: MailboxMetaData,
+                            mailboxSession: MailboxSession,
+                            allMailboxesMetadata: Seq[MailboxMetaData],
+                            quotaLoader: QuotaLoader) extends MailboxConstructionOrder {
```

Comment:
case classes are meant to be datastructures, you can't hold a service, use a `class` for that please

Reply:
after reading the file commit: you should probably have a smart-constructor for all the logic and a case class with the computed values. It would also prevent this ugly private for fields.

Reply:
How would you do a builder with Scala? I tried to look at it last week, I saw different ways of doing it, not sure what would be the best approach